common/src/main/java/io/pravega/common/concurrent/AsyncSemaphore.java

```diff
@@ -0,0 +1,220 @@
+ * number of credits, subject to a total number of credits being available. Each task's successful execution will "borrow"
+ * its share of credits, and a task cannot execute if the number of credits available is insufficient. Credits may be
+ * restored externally using the {@link #release} method.
+ *
+ * This is similar to {@link java.util.concurrent.Semaphore}, except that this class allows for asynchronous processing
+ * and each task can request an arbitrary numbe of credits. It can be useful in solving problems making use of the
+ * Leaky Bucket Algorithm (https://en.wikipedia.org/wiki/Leaky_bucket).
+ */
+@ThreadSafe
+public class AsyncSemaphore implements AutoCloseable {
+    //region Members
+
+    private final int totalCredits;
+    @GuardedBy("queue")
+    private int usedCredits;
+    @GuardedBy("queue")
+    private final ArrayDeque<DelayedTask> queue;
+    @GuardedBy("queue")
+    private boolean closed;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link AsyncSemaphore} class.
+     *
+     * @param totalCredits Total number of available credits.
+     * @param usedCredits  Initial number of used credits.
+     */
+    public AsyncSemaphore(int totalCredits, int usedCredits) {
+        Preconditions.checkArgument(totalCredits > 0, "totalCredits must be a positive integer");
+        Preconditions.checkArgument(usedCredits >= 0, "usedCredits must be a non-negative integer");
+        this.totalCredits = totalCredits;
+        this.usedCredits = usedCredits;
+        this.queue = new ArrayDeque<>();
+        this.closed = false;
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        List<DelayedTask> toCancel = null;
+        synchronized (this.queue) {
+            if (!this.closed) {
+                toCancel = new ArrayList<>(this.queue);
+                this.queue.clear();
+                this.usedCredits = 0;
+                this.closed = true;
+            }
+        }
+
+        if (toCancel != null) {
+            toCancel.forEach(q -> q.result.cancel(true));
+        }
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Executes the given task which requires the given number of credits.
+     *
+     * If there are sufficient credits available for this task to run, it will be invoked synchronously and the returned
+     * result is directly provided by the given task.
+     *
+     * If there are insufficient credits available for this task to run, it will be queued up and executed when credits
+     * become availble. There is no prioritization of queued tasks - they are triggered in the order in which they
+     * are queued up.
+     *
+     * A task will allocate the requested credits when it is triggered. If the task fails (synchronously or asynchronously),
+     * then the requested credits are automatically released back into the pool. If the task succeeds, the credits will
```

Comment:
The point of this is to keep track of unindexed table entries. When we accept new entries, we request credits. The successful completion of the "add new entry" means they are now in the system and are using up resources. The background indexer will process these and it will release credits as it makes progress.

If an entry is not accepted then it will not take up any resources so there it will not need its credits; those will be automatically added back to the pool.

Reply:
So you want the same Javadoc both at the class level and this method level? Someone using this should read and understand both, and I think the text here is sufficiently descriptive of its behavior.

Reply:
The docs here are sufficient to explain what this method does. But it may be useful to have an example on the top of the class to help a reader understand if the class is a good fit for their requirements.